### ISSUE TITLE
Likert-Scale, Pseudomarkers, etc.

### ISSUE THREAD

**Issue by lmhasse** [jtextor/immunet#1]
Hi,

I'm currently trying to run your pipeline on my own data with a new training suited for my data. I've tried it many times now, but I'm not quite happy with my results, but I think that's because the program isn't fitted to my data. It only has 4 markers instead of 6, but I already managed to change the 6 pseudomarkers to 4 etc. 

But during analyzing your code, I collected a few questions:

Can you explain to me how to interpret the Likert-Scale in annotations.json? 
I ran the training without the scale, but I guess it would be more accurate with information about the background.

Also, is there any way to get the annotation tool mentioned in the paper? I currently do it using ImageJ and then using a program I wrote to reformat it to your annotations format.

Why is the scale a list of length 6 when there's only five possible values (positive, probably positive, ...)? I expected the scale to be true in only one value, but after executing utils.py line 21 `return (np.array(c["positivity"][1:6]) - 1) / 4.` on the annotations.json , more than one value can be true. 

Then in train.py, for the `model_for_training` in `run_training`, why is the input shape fixed to `31*2-1`? Doesn't that exclude everything that isn't in that 63x63 window and ignore the annotations one made outside that window?

And last but not least: could you give me a brief explanation about what `phenotype()`, `phenotyped()` and `background()` are doing? 

Thank you so much in advance for any help.

Best,
Lucie

**Comment by EvgeniyaMartynova** [comment]
Hi lmhasse,

I'm sorry for a late reply, but I hope it can still be useful for you or others. And thank you for your questions.

**Likert-Scale in annotations.json**
In our annotations each cell has position on an image and positivity, which is an array that specify expression of phenotype markers used for training with  the Likert-Scale. The Likert-Scale express the degree of marker expression on a scale from 1 to 5, where 1 means no expression and 5 clear expression. 
1: negative
2: probably negative
3: unclear
4: probably positive
5: positive

We chose to use Likert-Scale because expression of some markers, like CD45RO, can be more like a spectrum for different cells. Hence, usage of Likert-Scale helps to express varied intensity for such markers and also certainty of annotator that a marker is expressed.

For instance for panel with DAPI, CD3, FOXP3, CD20, CD45RO, CD8 `"positivity": [5, 5, 1, 1, 1, 5]` means that it is a CD3+ CD8+ T cell.  And `"positivity": [5, 5, 1, 1, 4, 1]` means CD3+ CD45RO+ T cell with weaker but distinguishable CD45RO signal.

In practice we found that the perfromance is better with more certain values: 1 and 5, but it depends on your research question. If you want to accurately detect cells of different phenotype, I would advise using 1 or 5 for all markers, but if you want to get marker expression level from the model output, you should use other Likert-Scale values.

You can find more information about it in the paper preprint, in particular figures 2C and 3F.

You are saying that you run training without Likert-Scale, then I would guess that it is one of main reasons of poorer perfromance on your data.

**Annotation tool**

Unfortunately this software not available outside of our institute and it will be quite a lot of work to set up a way to distribute it. At the moment there are no clear plans about it.

> Why is the scale a list of length 6 when there's only five possible values (positive, probably positive, ...)? I expected the scale to be true in only one value, but after executing utils.py line 21 return (np.array(c["positivity"][1:6]) - 1) / 4. on the annotations.json , more than one value can be true.

I think it should be clear after explanation about the Likert-Scale. Positivity is a list of 6 because we use 6 markers channels to train the model and values in a list correspond to expression of each marker, e.g DAPI:5, CD3:5, FOXP3:1, CD20:1, CD45RO:1, CD8:1. More detailed explanation can be found in the pre-print.

> Then in train.py, for the model_for_training in run_training, why is the input shape fixed to 31*2-1? Doesn't that exclude everything that isn't in that 63x63 window and ignore the annotations one made outside that window?

Training is indeed done not on whole images but on patches of size  63x63 (you can experiment with size of patches).
During training  patches of size 63x63 are sampled in a way that regions of images with annotations are sufficiently represented. It is done in `make_training_data` function.

`phenotype(), phenotyped(), background()`

We have 2 types of annotations: _foreground_ and _background_. Foreground are cells that we want a model to detect, e.g. lymphocytes. Background annotations are done to help the model to understand that there is no cell of interest at a particular place in tricky cases (No cell, Neural structure) or that there is a cell we are not interested in (e.g Tumor cell, Other cell).

Hence `background ()` returns true of annotation type is  _background_.

In annotations.json you can notice that some annotations do not have positivity, e.g `{"type": "B cell", "x": 54, "y": 112}`. This is because for our example panel we know that B cells can only have positivity [5, 1, 1, 5, 1, 1]. On the contrary T cells can have different phenotypes: CD3+, CD3+ CD8+, CD3+ FOXP3+, CD3+ CD45RO+ and their annotations should have positivity to be used in training: `{"type": "T cell", "x": 13, "y": 143, "positivity": [3, 5, 5, 1, 5, 1]}`.

`phenotyped()` checks if an annotation can be used for training. This is the case type is among those for which there is only one positivity option or positivity is assigned in annotation.

`phenotype()` extracts a cell phenotype in a format that is necessary to make labels to train a model. For B cells we take its known phenotype even if positivity is annotated, for T cells, Dendritic cells, NK cells, NKT cells we take annotated positivity and transform it to scale from 0 to 1 and other annotations are background which we do not want to detect, that is why we return [0,0,0,0,0] for them.

We take only 5 last components of positivity because positivity assigned to DAPI is not used to generate labels.

I hope this explanation makes it clearer how to use the model for custom data, however, I see that it is rather tricky because annotation format is not documented and  `phenotype(), phenotyped(), background()` functions are highly specific for the panel used in the paper. We should make it more easy to adjust the code in future.

**Comment by lmhasse** [comment]
Hi Evgeniya,

Thank you for your response, that made it definitely clearer to me. In our lab we are currently trying to run your code on our data (only three markers to test) and we finally managed to get some good results, also I did include the likert scale. I also changed the code to be more flexible in the sense of changeable parameters and developed another way to annotate cells via imagej and some code I wrote, if you're interested in that let me know.

There are some other questions that came up to me during that process.We tried changing the parameter cell radius and could see that by enlarging the radius to 7/10/13 pixels, there were much more cells detected than by lowering it to 3 pixels (here predicted cells were much more fitting to the actual images). That somehow looks contrary to me. I would guess there's less cells if each cell has a higher radius and also the higher radius (around 10 pixels) is much more fitting to the average radius I detected in our tiff images. Could you explain that to me? I hope I explained clearly enough what I meant.

Still, even with a radius of 3 px, there's too many cells detected, especially in the membrane area of a cell. Could that be because of too little training data? 

Then there's some questions I have on the tiff images you give as input in the training. Are those already spectrally unmixed? And how are the tiles generated? I would appreciate a guide on where they're generated from.

Thank you for your help!

**Comment by EvgeniyaMartynova** [comment]
Hi lmhasse,

The version of the code we have now internally is quite different from what is published in this repo and some work has been done to make code more flexible, modular and easier to run. Also there are plans to migrate it to TF2. I'm not sure if we update the code when the paper is ready for publication, but if we see that community is interested in usage of a pipeline, perhaps we can release a more mature version in future. I will discuss it with my colleagues. Thank you for your suggestions!

I'm not sure why more cells are found if you increase a radius, but I have an idea. I guess, you have changed `cell_radius` parameter in `make_training_data()` which is then passed to `extract_labels()`. If so, it enlarges radius of distance and phenotype maps that are used as labels for training. Then, detection of cell is done with this line:
```
cell_locations = [(int(x[0]), int(x[1]))
            for x in blob_log(dist, min_sigma=3, max_sigma=5, threshold=log_threshold)]
```

This is a method for blobs detection with Laplacian of Gaussian filter. `min_sigma` and  `max_sigma` parameters specify minimum and maximum standard deviation of a Gaussian kernel used for blob detection. Hence to detect larger cells you need to increase both. When you enlarge disks that are generated around cell centers in labels for training, the network will output similarly large areas around each found cell, but then since `max_sigma ` is small, I'd guess that it will lead to detection of Fps by `blob_log`.

It is difficult to say why too many cells are detected even with a small radius without looking at the data. How much annotations do you have?

The images are obtained with Vectra 3 system and these tiff files are spectrally unmixed. This is done with inForm software provided with Vectra 3 system.

**Comment by lmhasse** [comment]
Hi Evgeniya,

I ran the training with around 2000 annotations. Now I'm running another one with around 50000.

I changed the code part to `...min_sigma=cell_radius-2, max_sigma=cell_radius...` and the predictions look much better now, higher radius gives less cells. Thank you! 

Do you have any updates concerning the input images? In my new training I'm also using spectrally unmixed images from inForm.

**Comment by EvgeniyaMartynova** [comment]
Hi lmhasse,

Great that I was able to help 😊.

Yes the image that come from the inForm analysis and used to train the model are indeed spectrally unmixed. I've updated my previous answer.
